### PR TITLE
Property inheritance ADR

### PR DESCRIPTION
**Pull request by jongeorge1**

### CODE REVIEW

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+## Context
+
+Corvus.Tenancy supports a hierarchy of tenants. There are two things that this can be used for:
+
+Firstly, we can control on a per-tenant basis where the children of that tenant are stored. For example, with two sibling tenants A and B (i.e. tenants that are children of the same parent tenant), the children of A can be stored in a completely different location to the children of B. By default, this will be a separate container in the same storage account, but it could be a completely separate storage account.
+
+Secondly, it can be used to enable better organisation of tenants by using parent tenants to group related tenants together.
+
+One of of the functions of tenants is to hold client-specific configuration for the applications that a client is using. An example would be for a client using the Workflow service, their tenant will contain two pieces of storage information, one for Workflows and one for Workflow instances. This configuration is stored in a collection of key-value pairs attached to the tenant.
```
Comment (mwadams) [inline docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md]: Should be a full stop not a comma before "their".

**Comment by jongeorge1** [comment]
> It also feels like their is a potential security issue with blanket inheritance. What if there are some parent properties you should not be able to inherit. While secrets _should_ not be stored in property bag, there may be something you don't want a child to know about the parent's properties and you could "probe" for settings.

**Comment by jongeorge1** [comment]
> It also feels like their is a potential security issue with blanket inheritance...

I mention this in the last paragraph of the `Decision` section. Are you suggesting it should be more prominent/explicit?

**Comment by mwadams** [comment]
Yes, I think I meant calling out an example of the mechanism but which this could leak (e.g. probing) and that it applies to any parent child tenant relationship, not just "dependent services".

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Status
+
+Proposed
+
+## Context
+
+Corvus.Tenancy supports a hierarchy of tenants. There are two things that this can be used for:
+
+Firstly, we can control on a per-tenant basis where the children of that tenant are stored. For example, with two sibling tenants A and B (i.e. tenants that are children of the same parent tenant), the children of A can be stored in a completely different location to the children of B. By default, this will be a separate container in the same storage account, but it could be a completely separate storage account.
```
Comment (idg10) [inline docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:11]: >where the children of that tenant are stored

I'm wondering if we might need to be more precise, because it's awfully easy to get confused between tenanted storage and tenant storage.

I tripped up on this sentence because I think of the "tenant" as being an entity that may be entirely outside of the system. Contoso is a tenant. Litware is a tenant. (Some tenants happen to be internal too; `Marain.Workflow` is a tenant, `Marain.Operations` is a tenant. But even here, the tenant is still an entity that is essentially external to `Corvus.Tenancy`.)

So I am uncomfortable saying that an `ITenant` instance _is_ the tenant. This may sounds like splitting hairs, but in this case, I think there is an important distinction to be made between the tenant and the object that represents the tenant. And the main reason for this is that the way we use this in Marain, we may actually have multiple different `ITenant` instances all corresponding to the same actual tenant, because of the on-behalf-of tenants we create when one service uses another service as part of its implementation.

So there isn't necessarily a one-to-one mapping between tenants of a multi-tenants service and `ITenant` instances.

On the other hand, that's a detail of how we use tenancy in Marain, so we don't want to bake that particular set of usage conventions into `Corvus.Tenancy`. So possibly we don't need to change it. But there is something about the phrase "where the children of that tenant are stored" that doesn't quite sit right with me, because tenant is invariably an overloaded term. What we actually store is a) properties, and b) a list of children.

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+Secondly, it can be used to enable better organisation of tenants by using parent tenants to group related tenants together.
+
+One of of the functions of tenants is to hold client-specific configuration for the applications that a client is using. An example would be for a client using the Workflow service. Their tenant will contain two pieces of storage information, one for Workflows and one for Workflow instances. This configuration is stored in a collection of key-value pairs attached to the tenant.
+
+It is possible for tenants to have child tenants in the hierarchy. If a tenant that uses the Workflow service has children, they may also need to use the Workflow service. In this case we have a choice: we can decide that we will allow the workflow storage configuration from a tenant to be inherited by its children, or we can require each tenant to contain all of it's own configuration.
+
+## Decision
+
+We have determined that we will not allow properties of tenants to be inherited by their children.
```
Comment (idg10) [inline docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:21]: "allow" seems slightly like the wrong word here. `Corvus.Tenancy` won't prevent me from writing something that deliberately copies all the properties down when creating new tenants, or when modifying existing ones, so it's not really disallowing it. It's just providing me with absolutely no support.

So I think I'd rephrase this along the lines that `Corvus.Tenancy` does not provide any means by which properties of tenants are inherited by their children. (It doesn't do it. But equally it wouldn't stop you from doing it if you don't mind doing all the relevant work. E.g., an implementation of `ITenantProvider` could, if it wanted, automatically copy the storage settings being used for the tenancy storage into child tenants by default, and `Corvus.Tenancy` allows this in the sense that it won't stop it.)

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+In the case when hierarchy is used for organisational purposes, inheritance is not relevant; parent tenants are there solely to group their children and configuration for the parent tenant is irrelevant, as it does not exist to be used as a tenant in it's own right.
+
+In the case where hierarchy represents a genuine parent-child relationship there are many potential reasons for this, and the goal of the project is not to dictate specific use cases. However, in making the decision not to implement property inheritance it is only necessary to find a use case where it is not desirable.
+
+Our use case here is a PaaS product providing multiple services - endjin's Marain platform. This platform contains several base services - Tenancy, Workflow, Operations and Claims, which can be licenced by clients.
+
+A client may choose to use these services to build their own platform, and use Marain's tenancy service to provide their own platform services to their own customers. In this case, the client's customers will be represented by child tenants of it's own tenant.
```
Comment (idg10) [inline docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:31]: "it's" should be "its".

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md (hunk 3 above)
Comment (idg10) [reply docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:21]: (Indeed, you go on to say that later. So "not allow" isn't quite right.)

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md (hunk 2 above)
Comment (jongeorge1) [reply docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:11]: Have reworded slightly to try and make this distinction.

File: docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+## Consequences
+
+Whilst not providing inheritance as standard reduces the risk of leaking configuration between parent and child tenants, it is possible to envisage scenarios where inheritance would be desirable. In these scenarios it will instead be necessary to manually copy configuration from parents to children, and potentially to maintain multiple copies of it.
+
+In this situations, client applications can simulate inheritance by automating the duplication of properties and by maintaining additional properties that indicate when values are effectively "inherited" vs when they have been explicitly set on the child tenants.
```
Comment (idg10) [inline docs/adr/0001-tenants-will-not-inherit-their-parents-properties.md:41]: this -> these
